### ISSUE TITLE
`builtins.fetchTree` and `builtins.fetchTarball` crash for tarballs with bogus `u-x` directory perms.

### ISSUE THREAD

**Issue by aameen-tulip** [NixOS/nix#7175]
**Describe the bug**

The builtin fetchers crash and burn for tarballs that carry evil permission bits on directories. I'm sure this is a side effect of running `tar c` on a wonky system or something, but if a tarball has a directory with perms like this you get a permissions failure during the post-unpack check routine in `libutil/tarfile.cc`.
```
~/repos/nix % tar tfv ./tarball.tgz
drw-rw-rw- 0/0               0 2020-02-18 03:50 package
-rw-rw-rw- 0/0             297 2020-02-18 03:50 package/index.d.ts
-rw-rw-rw- 0/0            1920 2020-02-18 03:50 package/index.js
-rw-rw-rw- 0/0            1092 2020-01-31 04:31 package/LICENSE
-rw-rw-rw- 0/0             937 2020-02-18 03:51 package/package.json
-rw-rw-rw- 0/0             713 2020-02-18 03:50 package/README.md
```

People shouldn't be publishing releases with permissions like this; but unfortunately it effects ~1/3,000 tarballs in the NPM registry. This really threw a wrench in a big Node.js+Flakes effort I've been working on.

I'm not sure if the `umask` settings of the user or daemon or builder are the ones being applied, but I suspect that it's using the `nix-daemon` perms as `root` because  otherwise `libarchive` is supposed to respect `umask`. Notice in the perms above the group and user IDs are set to `0/0` which I expect is relevant.

**Steps To Reproduce**

```
nix eval --impure --expr 'builtins.fetchTarball { url = "https://registry.npmjs.org/char-regex/-/char-regex-1.0.2.tgz"; }'
error (ignored): error: end of string reached
error: getting status of '/private/tmp/nix-42779-0/package/LICENSE': Permission denied
# or
nix eval --impure --expr 'builtins.fetchTree { type = "tarball"; url = "https://registry.npmjs.org/char-regex/-/char-regex-1.0.2.tgz"; }'
error (ignored): error: end of string reached
error: getting status of '/private/tmp/nix-42786-0/package/LICENSE': Permission denied
```

**Expected behavior**

untarred dirs should be readable by the user who invoked `nix`.

**`nix-env --version` output**
nix-env (Nix) 2.10.3

**Additional context**

I'm running on Darwin, but IIRC this same issue effected my Linux boxes. They're powered off now so I can confirm that later.

I tried a bunch of different `ARCHIVE_EXTRACT_*` combinations with no success but I'm far from a `libarchive` expert. A simple fix might just be to reset `u+x` before attempting to read.

I'm happy to help resolve this but I'm hesitant to PR anything effecting permissions without a more expert set of eyes on it.

**Comment by yorickvP** [comment]
I think we should remove  `ARCHIVE_EXTRACT_PERM` and  `ARCHIVE_EXTRACT_FFLAGS`

**Comment by edolstra** [comment]
But we do need to preserve the executable bit on regular files (the only permission Nix cares about).

**Comment by aakropotkin** [comment]
This is a naive question but could we just ignore directory entries' executable/read bits?

Out of curiosity do you know which users' umask is being applied? I'm guessing the daemon user's so maybe tweaking it to use the caller's ( risks seemingly spurious cache misses probably )?

Even if I had a well hidden config option similar to `NIX_CURL_FLAGS` to ignore directory entries I'd be covered. Or a `fetch(Tarball|Tree) { umask = ...; }`.

**Comment by aakropotkin** [comment]
Just wanted to report back and say that this is fixed in Nix 2.12:
```
$ nix --version
nix (Nix) 2.12.0

$ nix eval --impure --expr 'builtins.fetchTarball { url = "https://registry.npmjs.org/char-regex/-/char-regex-1.0.2.tgz"; }';
"/nix/store/4v2zs1dnzf9ih16k2fnpvx2b7pnq2ml1-source"
```
